### DISCUSSION TITLE
Selection of events at the reco and parton level

### DISCUSSION THREAD

**Discussion by kanhaiya-gupta** [MadAnalysis/madanalysis5#75] [General]
Dear Experts,

I am trying to match the events at the reco level and parton level using Madanalysis5 in the expert mode. I simulated 10K events using MadGraph for the pp->Zh, Z->ll, and h-> bb. But I have nearly 60k Z boson events at the parton (truth level). Is there any way to select the truth events for which reco level events are present (Delphes output)?

In the Geant4 , there are eventNumber and runNumber of both reco level and truth level. So there is a possibility to select the events having eventNumber and runNumber at both reco and truth level.

I Hope, I made you understand my question.
[Event_Number.pdf](https://github.com/MadAnalysis/madanalysis5/files/8605849/Event_Number.pdf)
[M_Z.pdf](https://github.com/MadAnalysis/madanalysis5/files/8605850/M_Z.pdf)

Thanking you
Kanhaiya Gupta

**Comment by jackaraz** [comment]
Hi @kanhaiya-gupta, I'm not sure if I understand what you want to do. You can count the events and get the event numbers from there if you want to do something with the events. I haven't used Geant4, so I don't know what it does. If you use [SFS](https://arxiv.org/abs/2006.09387) through expert mode instead of Delphes, you can access all detector level and MC level information for all the events in your sample.

**Reply by kanhaiya-gupta** [reply]
Thank you for your answer. 
I wanted to plot the 2D histogram between the detector level and MC level events. As you can see the number is different 10K for detector level and 60K for MC level. 

So, I wanted to select the same MC level event for which detector level event is present.

**Reply by jackaraz** [reply]
Depending on your Delphes construction you might have MC level information in `event.mc()->particles()`. This assumes that `GenParticle` information has been written in your ROOT file.

**Reply by kanhaiya-gupta** [reply]
I am sorry, I don't think, I get it. This is how I select the MC-level events.

for(unsigned int i=0; i<event.mc()->particles().size(); i++)
  {
    const MCParticleFormat *prt = &(event.mc()->particles()[i]);
   double Z_mass = prt->m();
   int z_id = prt->pdgid();

    if(z_id == 23.){
       myHisto5->Fill(Z_mass);
   }

  }


But the events number at the detector level and MC level are different. I have attached the plots.  My question is: is there any way to select these events and take only MC level event (60k) for which detector level (10k) is present


[Event_Number.pdf](https://github.com/MadAnalysis/madanalysis5/files/8609235/Event_Number.pdf)
[M_Z.pdf](https://github.com/MadAnalysis/madanalysis5/files/8609236/M_Z.pdf)

**Reply by jackaraz** [reply]
Hi @kanhaiya-gupta event number can not be different but the particles might be different, assuming that delphes has been executed properly. Each event has a `GenParticle` input and reco input. Gen particle includes all the particles available in the HEPMC file and reco corresponds to the FS particles that are reconstructed with detector effects. I don't understand what you mean when you say 60K MC events and 10K reco events. Delphes does not remove events from your sample. Is there something else you applied out of default Delphes construction?

**Reply by kanhaiya-gupta** [reply]
Thank you.

It seems that I misunderstood something. Let me explain what I did and where I am confused. 

I simulated 10k events using MadGraph. So, I can see there are 10k entries in the histogram Event_Number.pdf or Event.size histogram. There are entries in the particles.mass leaf. So I used the code below to select the Z boson. The number of entries in the histogram seems to be higher than the simulated one.  

for(unsigned int i=0; i<event.mc()->particles().size(); i++)
{
const MCParticleFormat *prt = &(event.mc()->particles()[i]);
double Z_mass = prt->m();
int z_id = prt->pdgid();

if(z_id == 23.){
   myHisto5->Fill(Z_mass);
}

}

You mean that the event number that is shown in the Delphes tree is same for both detector level events and MC level events. If it is the case, then it solves my problem.  I am a bit confused because the number of entries in the shown histogram (M_Z) is quite high.

**Reply by jackaraz** [reply]
Hi @kanhaiya-gupta, in your HEPMC file, there might be multiple entries for Z, and Delphes might be just storing all the raw data (you will need to check Delphes manual for details). I suggest printing all the information about each Z boson in an event to see where they are coming from. Then you can write a more detailed if statement to choose the Z you need. You probably need to take the Z originating from the hard process. You can do so by checking the status ID, but the status ID can vary with respect to the generator you are using. For that, you will need to check the respective manual of the generator that you used for the particle shower.

**Reply by kanhaiya-gupta** [reply]
Thanks a lot. This is very helpful. :)

**Reply by kanhaiya-gupta** [reply]
Hello,

I was wondering if it would be possible to get the information about the branch named Event.Number of the Delphes tree using Madanalysis5 in export mode. Something like event.mc()->number() or event.mc()->Number(). 

But I can see that this is not defined in the manual.  Any suggestions to implement this will be very helpful.

**Reply by jackaraz** [reply]
Hi @kanhaiya-gupta, there is no such accessor. The easiest method is to count events in your analysis execution. And the more laborious way is to implement the method that you like in [here](https://github.com/MadAnalysis/madanalysis5/blob/main/tools/SampleAnalyzer/Interfaces/delphes/DelphesTreeReader.cpp) to extract the information and have an accessor [here](https://github.com/MadAnalysis/madanalysis5/blob/main/tools/SampleAnalyzer/Commons/DataFormat/MCEventFormat.h).

**Reply by kanhaiya-gupta** [reply]
Thank you very much.

**Reply by kanhaiya-gupta** [reply]
There seems to be a problem regarding data types with implementation in MCEventFormat.h. When I define Event.Number as MAunit64 or MAint64, all histograms are empty. If I do as MAint32 or MAunit32, event_number is incorrectly filled whereas others are correctly filled. The actual data type of Event.Number branch is Long64_t.

**Reply by jackaraz** [reply]
Hi @kanhaiya-gupta, I don't know what you are doing, so I can not help. If you are extending Ma5 usage with some Delphes functionality, we can not help since none of us is an official Delphes developer. So for this, you will need to find a solution to your issue.
